tiledb/sm/filter/bitsort_filter.cc

```diff
@@ -0,0 +1,450 @@
+    input->advance_offset(part_size);
+  }
+
+  // Output metadata is a view on the input metadata, skipping what was used
+  // by this filter.
+  auto md_offset = input_metadata->offset();
+  RETURN_NOT_OK(output_metadata->append_view(
+      input_metadata, md_offset, input_metadata->size() - md_offset));
+
+  return Status::Ok();
+}
+
+template <typename AttrType>
+Status BitSortFilter::unsort_part(
+    const std::vector<uint64_t> &cell_pos, ConstBuffer* input_buffer, Buffer* output_buffer) const {
+  // Read in data.
+  uint32_t buff_size = input_buffer->size();
+  assert(buff_size % sizeof(AttrType) == 0);
+  uint32_t num_elems_in_part = buff_size / sizeof(AttrType);
+
+  if (num_elems_in_part == 0) {
+    return Status::Ok();
+  }
+
+  const AttrType* input_array = static_cast<const AttrType*>(input_buffer->data());
+  AttrType* output_array = static_cast<AttrType*>(output_buffer->cur_data());
+
+  // Write in data in original order.
+  for (uint32_t i = 0; i < num_elems_in_part; ++i) {
+    output_array[i] = input_array[cell_pos[i]];
+  }
+
+  return Status::Ok();
+}
+
+Status BitSortFilter::run_reverse_dim_tiles(BitSortFilterMetadataType &bitsort_metadata, std::vector<uint64_t> &cell_pos) const {
+  std::vector<Tile*> &dim_tiles = bitsort_metadata.first.get();
+  auto cmp_fn = bitsort_metadata.second.get();
+
+  for (uint64_t i = 0; i < dim_tiles[0]->cell_num(); ++i) {
+    cell_pos.push_back(i);
+  }
+
+  // Sort the dimension data and determine the positions of the dimension tile data.
+  std::sort(cell_pos.begin(), cell_pos.end(), cmp_fn);
+
+  // Rewrite the individual tiles with the position vector.
+  for (auto *dim_tile : dim_tiles) {
+  // Since run_reverse_dim_tile is only moving around dimension data and 
+  // reordering it, the type becomes irrelevant and only the size of the type
+  // matters, so the code cases on the size of the type and passes in the
+  // corresponding integer type into a templated function.
+    Datatype tile_type = dim_tile->type();
+    switch (datatype_size(tile_type)) {
+      case sizeof(uint8_t): {
+        run_reverse_dim_tile<uint8_t>(dim_tile, cell_pos);
+      } break;
+      case sizeof(uint16_t): {
+        run_reverse_dim_tile<uint16_t>(dim_tile, cell_pos);
+      } break;
+      case sizeof(uint32_t): {
+        run_reverse_dim_tile<uint32_t>(dim_tile, cell_pos);
+      } break;
+      case sizeof(uint64_t): {
+        run_reverse_dim_tile<uint64_t>(dim_tile, cell_pos);
+      } break;
+      default: {
+        return Status_FilterError(
+          "BitSortFilter::run_reverse_dim_tiles: dimension datatype does not have an appropriate "
+          "size");
+      }
+    }
+  }
+
+  return Status::Ok();
+}
+
+template<typename DimType>
+Status BitSortFilter::run_reverse_dim_tile(Tile *dim_tile, std::vector<uint64_t> &cell_pos) const {
+  // Obtain the pointer to the data the code modifies.
+  uint64_t cell_pos_size = cell_pos.size();
+  std::vector<DimType> tile_data_vec(cell_pos_size);
```

Comment:
Let's get rid of tile_data_vec and write directly into the tile. This way we remove an extra allocation which will probably give us a good performance boost.